core/src/main/scala/pureconfig/ProductHint.scala

```diff
@@ -0,0 +1,40 @@
+package pureconfig
+
+/**
+ * A trait that can be implemented to customize how case classes are read from and written to a config.
+ *
+ * @tparam T the type of case class for which this hint applies
+ */
+trait ProductHint[T] {
+
+  /**
+   * The mapping used to map field names to config keys
+   */
+  def fieldMapping: ConfigFieldMapping
```

Comment:
I think we can leave the `ConfigFieldMapping` and the `apply` method of `ProductHint` as is, and delegate the call in the default `ProductHintImpl` to the `ConfigFieldMapping`, no? Something along the lines of (if the method in `ProductHint` would be called `configFieldName`):

```scala
private[pureconfig] case class ProductHintImpl[T](
    fieldMapping: ConfigFieldMapping,
    useDefaultArgs: Boolean,
    allowUnknownKeys: Boolean) extends ProductHint[T] {
  def configFieldName(typeFieldName: String) = fieldMapping(typeFieldName)
}
```